pkg/indexer/blockchain/rpcLogStreamer.go

```diff
@@ -0,0 +1,167 @@
+	// All the listeners
+	contractConfigs []contractConfig
+	logger          *zap.Logger
+	rpcUrl          string
+}
+
+func NewRpcLogStreamBuilder(rpcUrl string, logger *zap.Logger) *RpcLogStreamBuilder {
+	return &RpcLogStreamBuilder{rpcUrl: rpcUrl, logger: logger}
+}
+
+func (c *RpcLogStreamBuilder) ListenForContractEvent(fromBlock int, contractAddress common.Address, topics []common.Hash) <-chan types.Log {
+	eventChannel := make(chan types.Log)
+	c.contractConfigs = append(c.contractConfigs, contractConfig{fromBlock, contractAddress, topics, eventChannel})
+	return eventChannel
+}
+
+func (c *RpcLogStreamBuilder) Build() (*RpcLogStreamer, error) {
+	client, err := ethclient.Dial(c.rpcUrl)
+	if err != nil {
+		return nil, err
+	}
+	return NewRpcLogStreamer(client, c.logger, c.contractConfigs), nil
+}
+
+// Struct defining all the information required to filter events from logs
+type contractConfig struct {
+	fromBlock       int
+	contractAddress common.Address
+	topics          []common.Hash
+	channel         chan<- types.Log
+}
+
+/*
+*
+A RpcLogStreamer is a naive implementation of the ChainStreamer interface.
+It queries a remote blockchain node for log events to backfill history, and then streams new events,
+to get a complete history of events on a chain.
+*
+*/
+type RpcLogStreamer struct {
+	client     ChainClient
+	watchers   []contractConfig
+	ctx        context.Context
+	cancelFunc context.CancelFunc
+	logger     *zap.Logger
+}
+
+func NewRpcLogStreamer(client ChainClient, logger *zap.Logger, watchers []contractConfig) *RpcLogStreamer {
+	return &RpcLogStreamer{
+		client:   client,
+		watchers: watchers,
+		logger:   logger,
+	}
+}
+
+func (r *RpcLogStreamer) Start(ctx context.Context) error {
+	r.ctx, r.cancelFunc = context.WithCancel(ctx)
+
+	for _, watcher := range r.watchers {
+		go r.watchContract(watcher)
+	}
+	return nil
+}
+
+func (r *RpcLogStreamer) watchContract(watcher contractConfig) {
+	fromBlock := int(watcher.fromBlock)
+	logger := r.logger.With(zap.String("contractAddress", watcher.contractAddress.Hex()))
+	for {
+		select {
+		case <-r.ctx.Done():
+			logger.Info("Stopping watcher")
+			return
+		default:
+			logs, nextBlock, err := r.getNextPage(watcher, fromBlock)
+			if err != nil {
+				logger.Error("Error getting next page", zap.Int("fromBlock", fromBlock), zap.Error(err))
+				time.Sleep(ERROR_SLEEP_TIME)
+				continue
+			}
+
+			logger.Info("Got logs", zap.Int("numLogs", len(logs)), zap.Int("fromBlock", fromBlock))
+			for _, log := range logs {
+				watcher.channel <- log
+			}
+			if nextBlock != nil {
+				fromBlock = *nextBlock
+			}
+		}
+	}
+}
+
+func (r *RpcLogStreamer) getNextPage(config contractConfig, fromBlock int) (logs []types.Log, nextBlock *int, err error) {
+	highestBlock, err := r.client.BlockNumber(r.ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	highestBlockCanProcess := int(highestBlock) - LAG_FROM_HIGHEST_BLOCK
+	numOfBlocksToProcess := highestBlockCanProcess - fromBlock + 1
```

Comment:
Why is it +1 here? Is the fromBlock already processed, or not processed yet?

Reply:
It's the number of blocks you want to have in a single query. So even if you have `from: 5, to: 5` you are actually processing 1 block since it seems both the from and to are inclusive.